### PR TITLE
 Adding Snapshot Options to deal with Server Timestamps

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This adds to new deserialization options for pending server timestamps - returning the previous value or returning an estimated timestamp.

Port of https://github.com/firebase/firebase-ios-sdk/pull/537

### CODE REVIEW

File: packages/firestore/test/integration/api/server_timestamp.test.ts [hunk 1]
```diff
@@ -158,7 +200,177 @@ apiDescribe('Server Timestamps', persistence => {
...
+  it('can return previous value of different type', () => {
+    return withTestSetup(() => {
+      return writeInitialData()
+        .then(() => docRef.update(updateData))
+        .then(waitForLocalEvent)
+        .then(snapshot => verifyTimestampsUsePreviousValue(snapshot, null))
+        .then(waitForRemoteEvent)
+        .then(snapshot => {
+          verifyTimestampsAreResolved(snapshot);
+        })
```
Comment (mikelehen) [inline packages/firestore/test/integration/api/server_timestamp.test.ts]: Lines 245-251 are repeated from the previous test, right?  Why?  It's really hard to read tests that are long and repetitive. I can't figure out what's setup vs behavior being tested (comments can be used as a crutch, but you didn't do that either :-)).

My preferred way to do this test (in the spirit of DRY) would probably be to change writeInitialData() to accept an optional initial value for the when fields.

Then you can just do writeInitialData(/*whenValue=*/42).then(() => docRef.update(updateData)) and then wait for the events / do your validation.

File: packages/firestore/test/integration/api/server_timestamp.test.ts [hunk 2]
```diff
@@ -158,7 +200,177 @@ apiDescribe('Server Timestamps', persistence => {
...
+          docRef.update('a', firebase.firestore.FieldValue.serverTimestamp())
+        )
+        .then(waitForLocalEvent)
+        .then(snapshot => {
+          expect(snapshot.get('a', { serverTimestamps: 'none' })).to.equal(
+            null
+          );
+          expect(
+            snapshot.get('a', { serverTimestamps: 'estimate' })
+          ).to.be.instanceof(Date);
```
Comment (mikelehen) [inline packages/firestore/test/integration/api/server_timestamp.test.ts]: Are the none / estimate tests actually valuable?  We already test their behavior elsewhere. Else, please update the name of the test.

File: packages/firestore/test/integration/api/server_timestamp.test.ts [hunk 3]
```diff
@@ -158,7 +200,177 @@ apiDescribe('Server Timestamps', persistence => {
...
+          );
+        })
+        .then(waitForRemoteEvent)
+        .then(snapshot => {
+          expect(
+            snapshot.get('a', { serverTimestamps: 'none' })
+          ).to.be.instanceof(Date);
+          expect(
+            snapshot.get('a', { serverTimestamps: 'estimate' })
+          ).to.be.instanceof(Date);
```
Comment (mikelehen) [inline packages/firestore/test/integration/api/server_timestamp.test.ts]: Again, I'd prefer not throw in extra expectations just because it's convenient.  Testing the behavior of 'estimate' and 'previous' on actual resolved timestamps (aka normal Dates) is probably good, but should go elsewhere (either in the existing 'can return {previous,estimate}' tests or in a new test.

File: packages/firestore/test/integration/api/server_timestamp.test.ts [hunk 4]
```diff
@@ -158,7 +200,177 @@ apiDescribe('Server Timestamps', persistence => {
...
+
+      let promises: Promise<void>[] = [];
+
+      return writeInitialData()
+        .then(() => docRef.update(updateData))
+        .then(waitForLocalEvent)
+        .then(snapshot => {
+          verifyTimestampsUsePreviousValue(snapshot, null);
+          return waitForRemoteEvent();
+        })
```
Comment (mikelehen) [inline packages/firestore/test/integration/api/server_timestamp.test.ts]: Can you see if you can clean up the setup here too?

File: packages/firestore/test/integration/api/server_timestamp.test.ts [hunk 5]
```diff
@@ -158,7 +200,177 @@ apiDescribe('Server Timestamps', persistence => {
...
+        .then(waitForRemoteEvent)
+        .then(snapshot => {
+          return Promise.all(promises);
+        })
+        .then(() => {});
+    });
+  });
+
+  it('uses previous value from local mutation', () => {
```
Comment (mikelehen) [inline packages/firestore/test/integration/api/server_timestamp.test.ts]: It took me a couple minutes to figure out what was different about this test vs the previous one.  These tests really need less boilerplate and/or more comments. Or perhaps better would actually be to move these to unit tests rather than integration tests so you can more directly test the applyToLocalView() code without all of this extra setup / asynchrony / etc. in the way... :-/  The integration tests are for hitting all of the public surface area and sanity-checking behavior rather than covering all edge cases.

File: packages/firestore/src/api/database.ts [hunk 6]
```diff
@@ -982,6 +984,9 @@ export class DocumentReference implements firestore.DocumentReference {
   }
 }
 
+/** Options interface that can be provided to configure the deserialization of DocumentSnapshots. */
+export interface SnapshotOptions extends firestore.SnapshotOptions {}
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/api/database.ts]: We don't do this with the other options. The other options only have boolean values and we just access them directly (options.merge for example). I added this so I don't have to import the public type in field_value.ts.

File: packages/firestore/src/api/database.ts [hunk 7]
```diff
@@ -1702,6 +1723,25 @@ function validateSetOptions(
   return options;
 }
 
+function validateSnapshotOptions(
+  methodName: string,
+  options?: firestore.SnapshotOptions
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/api/database.ts]: That makes sense. Thanks for explaining the reasoning behind the validateSetOptions() signature.

File: packages/firestore/src/model/field_value.ts [hunk 8]
```diff
@@ -61,6 +62,34 @@ export enum TypeOrder {
...
+  constructor(readonly serverTimestampBehavior: ServerTimestampBehavior) {}
+
+  static fromSnapshotOptions(options: SnapshotOptions) {
+    switch (options.serverTimestamps) {
+      case 'estimate':
+        return new FieldValueOptions(ServerTimestampBehavior.Estimate);
+      case 'previous':
+        return new FieldValueOptions(ServerTimestampBehavior.Previous);
+      case 'none': // Fall-through intended.
+      default:
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/field_value.ts]: That works.

File: packages/firestore/src/model/field_value.ts [hunk 9]
```diff
@@ -80,7 +109,7 @@ export type FieldType = null | boolean | number | string | {};
 export abstract class FieldValue {
   readonly typeOrder: TypeOrder;
 
-  abstract value(): FieldType;
+  abstract value(options?: FieldValueOptions): FieldType | undefined;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/field_value.ts]: It looks like it is a mistake. ServerTimestamps return null but never undefined.

File: packages/firestore/src/model/field_value.ts [hunk 10]
```diff
@@ -319,18 +348,31 @@ export class TimestampValue extends FieldValue {
...
  * - With respect to other ServerTimestampValues, they sort by their
  *   localWriteTime.
  */
 export class ServerTimestampValue extends FieldValue {
   typeOrder = TypeOrder.TimestampValue;
 
-  constructor(readonly localWriteTime: Timestamp) {
+  constructor(
+    readonly localWriteTime: Timestamp,
+    readonly previousValue: FieldValue | null
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/field_value.ts]: Gil and I talked about overlays when we discussed this and hope that overlays make getting the previous value a lot easier.

File: packages/firestore/src/model/field_value.ts [hunk 11]
```diff
@@ -319,18 +348,31 @@ export class TimestampValue extends FieldValue {
...
   }
 
-  value(): null {
+  value(options?: FieldValueOptions): FieldType {
+    if (options) {
+      switch (options.serverTimestampBehavior) {
+        case ServerTimestampBehavior.Estimate:
+          return this.localWriteTime.toDate();
+        case ServerTimestampBehavior.Previous:
+          return this.previousValue ? this.previousValue.value(options) : null;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/field_value.ts]: It falls through to the function's main return value.

I changed this to an if/else to make this more obvious (and to not repeat the same error handling everywhere).

File: packages/firestore/src/model/mutation.ts [hunk 12]
```diff
@@ -226,13 +226,17 @@ export abstract class Mutation {
    *
    * @param maybeDoc The document to mutate. The input document can be null if
    *     the client has no knowledge of the pre-mutation state of the document.
+   * @param baseDoc The state of the document prior to this mutation batch. The
+   *     input document can be null if the client has no knowledge of the
+   *     pre-mutation state of the document.
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/mutation.ts]: `set({a:FieldValue.serverTimestamp()})` is a good example why this doesn't work. It will create two mutations: One that clears the map `a` and removes the existing value, and one that adds the timestamp. When we apply the timestamp, `previousValue` has already been cleared in `maybeDoc`.

File: packages/firestore/src/model/mutation.ts [hunk 13]
```diff
@@ -539,14 +549,27 @@ export class TransformMutation extends Mutation {
...
     const transformResults = [] as FieldValue[];
     for (const fieldTransform of this.fieldTransforms) {
       const transform = fieldTransform.transform;
       if (transform instanceof ServerTimestampTransform) {
-        transformResults.push(new ServerTimestampValue(localWriteTime));
+        let previousValue: FieldValue | null = null;
+
+        if (baseDoc instanceof Document) {
+          previousValue =
+            (baseDoc as Document).field(fieldTransform.field) || null;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/mutation.ts]: Oh, yes. I remember this now from a TS talk I watched I while ago :)

File: packages/firestore/src/typings/firestore.d.ts [hunk 14]
```diff
@@ -518,6 +518,29 @@ declare namespace firestore {
...
+   * (e.g. the desired behavior for server timestamps that have not yet been set
+   * to their final value).
+   */
+  export interface SnapshotOptions {
+    /**
+     * If set, controls the return value for server timestamps that have not yet
+     * been set to their final value.
+     *
+     * By specifying `estimate`, pending server timestamps return a local
+     * estimate. This estimate will likely differ from the final value and may
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/typings/firestore.d.ts]: Done.

File: packages/firestore/src/typings/firestore.d.ts [hunk 15]
```diff
@@ -518,6 +518,29 @@ declare namespace firestore {
...
+     * cause these values to change once the server result becomes available.
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/typings/firestore.d.ts]: Reworded.

File: packages/firestore/src/typings/firestore.d.ts [hunk 16]
```diff
@@ -518,6 +518,29 @@ declare namespace firestore {
...
+     *
+     * By specifying `previous`, these pending timestamps will instead retain
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/typings/firestore.d.ts]: Removed

File: packages/firestore/src/typings/firestore.d.ts [hunk 17]
```diff
@@ -518,6 +518,29 @@ declare namespace firestore {
...
+     * their previous value.
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/typings/firestore.d.ts]: Reworded.

File: packages/firestore/src/typings/firestore.d.ts [hunk 18]
```diff
@@ -518,6 +518,29 @@ declare namespace firestore {
...
+     *
+     * If omitted or set to `none`, `null` will be returned by default until the
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/typings/firestore.d.ts]: Fixed here and above.

File: packages/firestore/src/util/input_validation.ts [hunk 19]
```diff
@@ -169,6 +169,58 @@ export function validateNamedOptionalType(
   }
 }
 
+/**
+ * Validates that the provided named option equals one of the expected values.
+ */
+export function validateNamedPropertyEquals<T>(
+  functionName: string,
+  inputName: T,
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/util/input_validation.ts]: Fixed.

File: packages/firestore/src/util/input_validation.ts [hunk 20]
```diff
@@ -169,6 +169,58 @@ export function validateNamedOptionalType(
...
+  optionName: string,
+  input: T,
+  expected: T[]
+) {
+  const expectedDescription: string[] = [];
+
+  for (let val of expected) {
+    if (val === input) {
+      return true;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/util/input_validation.ts]: As written, it returns boolean. Changed to return void.

File: packages/firestore/test/integration/api/server_timestamp.test.ts [hunk 21]
```diff
@@ -63,35 +63,72 @@ apiDescribe('Server Timestamps', persistence => {
...
+      const when = prev.get('when');
+      expect(current.data({ serverTimestamps: 'previous' })).to.deep.equal(
+        expectedDataWithTimestamp(when)
       );
-
-      // Validate the rest of the document.
-      expect(remoteSnap.data()).to.deep.equal(expectedDataWithTimestamp(when));
-    });
+    }
   }
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/test/integration/api/server_timestamp.test.ts]: GIl asked me to split up the await logic from the verification. Unfortunately, this blows up the JS tests quite a bit. I'll keep the suggested cleanup in mind for when I need to ask for a favor :)

File: packages/firestore/test/integration/api/validation.test.ts [hunk 22]
```diff
@@ -241,6 +241,28 @@ apiDescribe('Validation:', persistence => {
...
+      .then(snapshot => {
+        expect(() => snapshot.get('test', { bad: true } as any)).to.throw(
+          `Unknown option 'bad' passed to function ` +
+            `DocumentSnapshot.get(). Available options: ` +
+            `serverTimestamps`
+        );
+        expect(() =>
+          snapshot.data({ serverTimestamps: 'foo' } as any)
+        ).to.throw(
+          `The serverTimestamps property of the options argument for DocumentSnapshot.data() is expected to be one of "estimate","previous","none",  but was "foo".`
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/test/integration/api/validation.test.ts]: Not using a space was my somewhat questionable workaround for not adding the "or" (I originally had a regexp-like thing in the error message that used `|` as a separator). Putting the values at the end nicely works around this problem. Thanks!

File: packages/firestore/test/integration/api/server_timestamp.test.ts (hunk 1 above)
Comment (schmidt-sebastian) [reply packages/firestore/test/integration/api/server_timestamp.test.ts]: I cleaned up the tests quite a bit and removed the parts that are exercised in other tests. I think this reduces the scope and duplication enough to not introduce further helpers.

File: packages/firestore/test/integration/api/server_timestamp.test.ts (hunk 5 above)
Comment (schmidt-sebastian) [reply packages/firestore/test/integration/api/server_timestamp.test.ts]: I took a quick look at testing this via the mutations batches earlier, and some of this testing can indeed be moved there by verifying `previousState` via `valueWithOptions`.
Since Server Timestamps were already mostly tested via these integration tests, I ended up shying away from this though. 
Before I start reworking this, I would like to know how strong you feel about it, as this change should then also be ported to the Android and iOS clients.

File: packages/firestore/test/integration/api/server_timestamp.test.ts (hunk 3 above)
Comment (schmidt-sebastian) [reply packages/firestore/test/integration/api/server_timestamp.test.ts]: Removed

File: packages/firestore/test/integration/api/server_timestamp.test.ts (hunk 4 above)
Comment (schmidt-sebastian) [reply packages/firestore/test/integration/api/server_timestamp.test.ts]: Done

**Comment by schmidt-sebastian** [comment]
I addressed all comments. I also pretended that Greg's disableNetwork PR already landed as this allowed me to clean up the tests further.

If you have time, I'd appreciate another round on this. Otherwise, we can wait until his PR is merged and then proceed with this review.

File: packages/firestore-types/index.d.ts [hunk 23]
```diff
@@ -565,20 +588,34 @@ export class DocumentSnapshot {
...
    */
-  data(): DocumentData | undefined;
+  data(options?: SnapshotOptions): DocumentData | undefined;
 
   /**
    * Retrieves the field specified by `fieldPath`. Returns 'undefined' if the
    * document or field doesn't exist.
    *
+   * By default, a `FieldValue.serverTimestamp()` that have not yet been set to
+   * its final value will be returned as `null`. You can override this by
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore-types/index.d.ts]: Fixed.

File: packages/firestore/src/model/field_value.ts [hunk 24]
```diff
@@ -349,7 +351,7 @@ export class TimestampValue extends FieldValue {
  *   the local view of a document. Therefore they do not need to be parsed or
  *   serialized.
  * - When evaluated locally (e.g. for snapshot.data()), they by default
- *   evaluate to `null`. This behavior can be configured by passing custom
+ *   evaluate to 'null'. This behavior can be configured by passing custom
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/src/model/field_value.ts]: Reverted.

File: packages/firestore/test/integration/api/validation.test.ts [hunk 25]
```diff
@@ -258,7 +258,7 @@ apiDescribe('Validation:', persistence => {
         expect(() =>
           snapshot.data({ serverTimestamps: 'foo' } as any)
         ).to.throw(
-          `The serverTimestamps property of the options argument for DocumentSnapshot.data() is expected to be one of "estimate","previous","none",  but was "foo".`
+          `Invalid value "foo" provided to function DocumentSnapshot.data() for option 'serverTimestamps'. Acceptable values: estimate, previous, none`
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply packages/firestore/test/integration/api/validation.test.ts]: I like the single quotes more. The double quotes come from JSON.stringify. In the interest of consistency, I persuaded myself to use double quotes everywhere.
